Tests/SessionTests.swift

```diff
@@ -1435,6 +1435,43 @@ class SessionTestCase: BaseTestCase {
         // Then
         assertErrorIsAFError(error) { XCTAssertTrue($0.isSessionDeinitializedError) }
     }
+
+    // MARK: Tests - Request Cancellation
+
+    func testThatSessionOnlyCallsResponseSerializerCompletionWhenCancellingInsideCompletion() {
+        // Given
+        let handler = RequestHandler()
+        let session = Session()
+
+        let expectation = self.expectation(description: "request should complete")
+        var response: DataResponse<Any>?
+        var completionCallCount = 0
+
+        // When
+        let request = session.request("https://httpbin.org/get", interceptor: handler)
+        request.validate()
+
+        request.responseJSON { resp in
+            request.cancel()
```

Comment:
It seems like we may want to more generally fix `cancel()` by preventing its actions when the `Request` has already finished. It may be out of scope for this fix, but it makes sense to me that we may want to add an additional stage to state, perhaps `.completed`, to indicate that a `Request` has finished and started calling its completion handlers.

Reply:
I agree and would advocate we consider that outside this particular PR. I think we need to spend some time thinking about the `resume`, `suspend`, and `cancel` APIs and the current safety around them. I think there are a lot of good additions to AF5 in this area, but I think they could still use another rev before we officially release.

Reply:
Sounds good!